### ISSUE TITLE
Stuck at this warning for >20 min: `UserWarning: Distributing <class 'dict'> object. This may take some time.`

### ISSUE THREAD

**Issue by joelostblom** [modin-project/modin#1572]
I am trying to use modin to speed up an image processing pipeline that uses pandas dataframes to store images in individual columns together with computed image properties in other columns. I am new to modin, and I was trying it as a drop-in replacement to pandas by changing the import statement to `import modin.pandas as pd`. However, when I run a function to compute a set of image properties which takes just a few minutes with standard pandas, modin gets stuck at the warning `UserWarning: Distributing <class 'dict'> object. This may take some time.` for at least 20 min (aborted manually).

I would like to try to narrow down this issue, but I don't know where to start looking and did not find anything on the issue tracker relating to this warning either. The function I am running is rather long, so not sure how helpful it would be to post it here, but it is essentially applying functions over the image columns either individually or together in permutations to compute properties of the images. I am already using joblib to parallelize, but for testing with modin I set the joblib loops to only use a single core.

I am sorry this is not more specific, is there anything that can be said about where to start looking for why I am getting stuck? Or is modin not a good fit for working with dataframes containing images?

I tried this with modin 0.7.3 installed from the conda-forge, and running with the dask backend. I tried installing ray and modin via pip, but could not get it to work (I could try more if you believe it would make a difference). I am on Arch Linux 5.6.15-arch1-1

**Comment by devin-petersohn** [comment]
Hi @joelostblom, thanks for the detailed question! I will explain some details about how Modin works first, then we can try to debug this performance.

When Modin takes in data, it partitions datasets in the way that makes them scalable. So when we get a dictionary object, we have to manually split the data to each partition and this can't (reliably) be done in parallel without making many copies because the data cannot be accessed by all of the Modin workers in parallel. There are still some tricks we can do here to make things marginally faster, but there is a limitation at the Python language level that will be hard to get around.

Ideally, we can ingest the data from a CSV file or some other format. Then all of the workers can access the shared filesystem and read in-parallel. This will be faster than having the data already in memory. Is it possible to read the data in from a file?

The joblib point is another challenge with Modin. If Modin only has 1 CPU it will basically be pandas + some communication overheads, so the more the better here. There are also known issues with Modin working with joblib (#672). I assume you were trying to manually parallelize a large job with pandas before, my recommendation here is to try Modin without joblib. 

Your type of workload is a good fit for what we want to do with Modin, so let's see if we can help your case more. A lot of the workarounds people used with pandas to fix certain things are counter-productive in Modin.

**Comment by joelostblom** [comment]
Thanks for your reply @devin-petersohn !

Your explanation makes a lot of sense to me. I did not previously understand that modin's parallelization depends is best applied during the reading of data, I thought it was generally when constructing a pandas dataframe. I think this makes sense with the error because I am reading tens of GBs of images into a dictionary that I am then converting to a dataframe and I guess this is where modin/python chokes.

I might try your suggestion with writing these images to a single CSV (probably as flattened arrays then) that I read into modin. I am a bit afraid that the tripled IO (read imgs, write csv, read csv) will negate much of the later performance benefit in terms of overall execution time. Still, I think it is an attractive option to allocate most of the processing time to the initial IO in order to speed up later parts of the pipeline, so I will give this a go (but it might be a little while before I can rewrite that part of the pipeline).

Does this makes sense to you, or do you see any more immediate way of scaling up the image? I have meant to look into storing images in an HDF5 store for other reasons, would this also be a suitable format for reading a dataframe into modin or is CSV preferred?

Once I have this initial dataframe of images, what happens to performance if I construct a second dataframe from the first one (e.g. through a reshape, subset, etc). Would I face similar problems as now since this second dataframe is not created directly from a read operation?

**Comment by devin-petersohn** [comment]
@joelostblom how are you initially reading the images? We have some extensible parsers that would potentially allow us to directly read those in parallel at first. Are you doing something with the dictionary or is it simply needed to create a dataframe? What format are you reading from?

The triple IO can potentially be problematic, and it would depend on what you're doing later. Depending on what you're doing with the dictionary we may be able to avoid this altogether. 

HDF5 support is not yet implemented in Modin. There are some limitations to the parent libraries we might depend on and haven't had the time to focus on writing a custom HDF5 engine. Anything we have implemented will work just fine, these can be found on the readme.

Constructing a new DataFrame from another will still be fast. The speed problem is coming from some structure that Modin doesn't control. Python dictionaries are more difficult to distribute because the Python executor that controls it must communicate it directly to other Python executors, so we are limited by how fast Python can loop over the data and send it. After all of this data is split up and distributed, Modin will run fast.

**Comment by joelostblom** [comment]
I use `imread` from scikit-image in a loop using joblib to paralellize over a list of all the file names to read in. The images are in tiff format, and I believe that scikit-image uses the tifffile library under the hood (these are microscopy images, so eventually I might switch to xarray instead of numpy since that seems to provide some conveniences , but this is more of a long term goal).

The only thing I am doing with the dictionary is storing some image metadata (short strings) that is extracted from the file names, but as long as I can store the file name and the image using the modin approach, I could extract this metadata as a separate step. I also downscale the images using scikit-learn right after reading them so that I don't need to store the full size image, but I imagine this would not be a problem to do with the approach you are suggesting (or possible as a separate step afterwards).

Let me know if you need more info or some sample code if it is not clear from the above. Excited to learn how I can speed this up using modin!

**Comment by joelostblom** [comment]
I should also say that the image processing I am doing relies on scikit-image and numpy so I would need to eventually have the images in a format that is compatible with those libraries (the scikit-image `imread` function returns a numpy array).

**Comment by devin-petersohn** [comment]
@joelostblom This helps quite a bit. Some code samples might help here. How many image files are you processing and what does the structure of a row or column look like in the dataframe?

Modin doesn't have a numpy API yet, so this might be a challenge at the current time. We have some plans in the works for this, but (as you can imagine) it is a big undertaking so we're making sure the design is right. We can convert a Modin dataframe to a numpy array at the moment, but it will not be distributed (yet).

We are also building out the extensibility component of the API and custom parsers. This would be a good test of that. There's no good high level API for these at the moment, so it will take some collaboration until we introduce a high level API for custom parsers and functions.

**Comment by joelostblom** [comment]
Sorry for the delayed reply @devin-petersohn, I have had some thesis deadlines to tend to over the weekend so got lost in that. 

It is unfortunate that there is no numpy API atm, then it sounds like modin might not be the tool I need for my image processing needs right now? In the meantime, I found https://github.com/nalepae/pandarallel which has been working great and fit well with my pipeline since it was mainly a few `.apply` call on the image columns that constituted the bottleneck for me currently. I believe modin would bring additional benefits when it starts working with numpy images, but for my current application I might stick with pandarallel since I already have a working solution for the performance issues I was experiencing.

In terms of your questions on the data, I am processing anything from 10 to 300-400 images, typically around 100. Then I process these with skimage to crop objects out of the images and measure the properties of these objects. So the initial dataframe has 1-4 big images per row (depending on how many channels we record in the microscope), but the final data frames with measurements has a single of the original images split out into multiple rows, one object per row. This would be the typical structure of the final data frame (truncated from around 20ish columns and hundreds of rows):

|    |   major_axis |   aspect_ratio |   perimeter |   area |   circularity | img1a | img1b |
|---:|-------------:|---------------:|------------:|-------:|--------------|---|---:|
|  0 |      79.103  |        1.14776 |     323.492 |   4222 |      0.859097 | 2D np.array | 2D np.array|
|  1 |      89.6788 |        1.26829 |     524.653 |   4836 |      0.765625 |2D np.array | 2D np.array|
|  2 |      78.4372 |        1.21918 |     352.599 |   3909 |      0.808969 |2D np.array | 2D np.array|

**Comment by mvashishtha** [comment]
I don't see a particular actionable task in this issue, so I'm going to close it. @joelostblom I hope you are using Modin. Please comment here or in a new issue if you would like help.
